core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,420 @@
+     * @param length
+     *          the number of bytes to be read from the given array; must be non-negative and no larger than row.length - offset
+     * @return this builder
+     */
+    Build visibility(final byte[] columnVisibility, int offset, int length);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the {@link Key}
+     * @return this builder
+     */
+    Build visibility(final Text columnVisibility);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the {@link Key}. The encoding must be UTF-8
+     * @return this builder
+     */
+    Build visibility(final CharSequence columnVisibility);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the {@link Key}
+     * @return this builder
+     */
+    Build visibility(final ColumnVisibility columnVisibility);
+  }
+
+  /**
+   * @since 1.9
+   */
+  static class KeyBuilderImpl implements RowStep, ColumnFamilyStep, ColumnQualifierStep,
+      ColumnVisibilityStep {
+
+    protected static final byte EMPTY_BYTES[] = new byte[0];
+
+    private final boolean copyBytes;
+    private byte[] row = EMPTY_BYTES;
+    private int rowOffset = 0;
+    private int rowLength = 0;
+    private byte[] family = EMPTY_BYTES;
+    private int familyOffset = 0;
+    private int familyLength = 0;
+    private byte[] qualifier = EMPTY_BYTES;
+    private int qualifierOffset = 0;
+    private int qualifierLength = 0;
+    private byte[] visibility = EMPTY_BYTES;
+    private int visibilityOffset = 0;
+    private int visibilityLength = 0;
+    private long timestamp = Long.MAX_VALUE;
+    private boolean deleted = false;
+
+    KeyBuilderImpl(boolean copyBytes) {
+      this.copyBytes = copyBytes;
+    }
+
+    private byte[] copyBytesIfNeeded(final byte[] bytes, int offset, int length) {
+      return Key.copyIfNeeded(bytes, offset, length, this.copyBytes);
+    }
+
+    @Override
+    public ColumnFamilyStep row(final byte[] row, int offset, int length) {
+      this.row = copyBytesIfNeeded(row, offset, length);
+      this.rowOffset = this.copyBytes ? 0 : offset;
+      this.rowLength = this.copyBytes ? this.row.length : length;
+      return this;
+    }
+
+    @Override
+    public ColumnFamilyStep row(final byte[] row) {
+      return row(row, 0, row.length);
+    }
+
+    @Override
+    public ColumnFamilyStep row(final Text row) {
+      return row(row.getBytes(), 0, row.getLength());
+    }
+
+    @Override
+    public ColumnFamilyStep row(final CharSequence row) {
+      return row(new Text(row.toString()));
```

Comment:
Done, now I'm using `CharBuffer` to encode any `CharSequence`. If the `CharSequence` is not UTF-8 coded, the `KeyBuilder` throws a `RuntimeException` containing a `CharacterCodingException`.